engine/frontend/src/component/log/Log.js

```diff
@@ -0,0 +1,178 @@
+const convert = new Convert();
+
+const Row = ({ data, index, setSize, windowWidth}) => {
+  const rowRef = useRef();
+  
+  useEffect(() => {
+    if (rowRef.current !== undefined) {
+      setSize(index, rowRef.current.getBoundingClientRect().height);
+    } 
+  }, [setSize, index, windowWidth]);
+
+  if (data[index] !== undefined && data[index].length !== 0) {
+    let txt = data[index]
+    if (hasAnsi(data[index])) {
+      const parsedAnsi = convert.toHtml(data[index])
+      txt = parse(parsedAnsi)
+    }
+    return (
+        <div
+          ref={rowRef}
+          style={styles.row}
+        >
+         {txt}
+        </div>
+      );
+  } 
+  return null;
+};
+
+export const Log = ({logs, fileName}) => {
+  const { onCopy, value:copyValue, setValue:setCopyValue, hasCopied } = useClipboard("");
+
+  const [initalRender, setInitialRender] = useState(false);
+  const listRef = useRef(null);
+  const sizeMap = useRef({});
+  const testRef = useRef({})
+  const containerRef = useRef({})
+  const [bottomMap, setMaybeBottom] = useState([0, 0])
+  const [previousOffset, setPreviousOffset] = useState(0)
+
+
+  const setSize = useCallback((index, size) => {
+    sizeMap.current = { ...sizeMap.current, [index]: size };
+    listRef.current.resetAfterIndex(index);
+  }, []);
+
+  const getSize = index => sizeMap.current[index] || 50;
+  const [windowWidth, windowHeight] = useWindowResize();
+
+  const [isBottom, setBottom] = useState(true)
+
+  useEffect(() => {
+    const logsWithoutAnsi = logs.map((log)=> {
+      return stripAnsi(log)
+    })
+
+    setCopyValue(logsWithoutAnsi.join("\n"))    
+    // automatically scroll the bottom if it;s at the bottom
+    if (logs.length > 0 && isBottom) {
+      scrollToBottom()
+    }
+
+  }, [logs]);
+  
+  const scrollToBottom = () => {
+    if (listRef.current) {
+      console.log("scroll to ", logs.length)
+      listRef.current.scrollToItem(logs.length - 1, "end");
+    }
+  }
+
+  const handleScroll = ({scrollOffset}) => {
+      // user manually scrolled up 
+      if (scrollOffset < previousOffset && testRef.current.clientHeight > containerRef.current.clientHeight) {
+        console.log("setting scroll to false!")
+        setBottom(false);
+      } 
+      setPreviousOffset(Math.floor(scrollOffset))
+  };
+
+  const handleToBottom = () => {
+    scrollToBottom()    
+    setBottom(true)
+  }
+
+  const handleItemsRendered = ({
+    overscanStartIndex,
+    overscanStopIndex,
+    visibleStartIndex,
+    visibleStopIndex}) => {
+      const maybeBottom = [overscanStopIndex, visibleStopIndex]
+      if (overscanStopIndex === logs.length - 1) {
+        setBottom(true)
+      }
+  }
+
+  const isItAtBottom = () => {
+    console.log(logs.length - 1, bottomMap[0])
+    return bottomMap[0] === logs.length - 1
+  }
+
+  const handleDownload = () => {
+    const logsWithoutAnsi = logs.map((log)=> {
+      return stripAnsi(log)
+    })
+    const logsAsText = logsWithoutAnsi.join("\n")
```

Comment:
@Peeeekay wondering if we can use a platform agnostic value?
https://stackoverflow.com/a/12989543